File: app/src/main/java/com/example/toucheventexplorer/OnScreenLogAdapter.java
package com.example.toucheventexplorer;

import android.text.SpannableString;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.List;

import javax.annotation.Nonnull;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

class OnScreenLogAdapter extends RecyclerView.Adapter {
    private final List<EventLogEntry> mlog;

    public OnScreenLogAdapter(@Nonnull List<EventLogEntry> log) {
        mlog = log;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        AppCompatTextView line = (AppCompatTextView) LayoutInflater.from(parent.getContext())
            .inflate(R.layout.log_entry, parent, false);
        return new LogViewHolder(line);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        LogViewHolder viewHolder = (LogViewHolder) holder;
        SpannableString ss = mlog.get(position).getSpannableString();
        viewHolder.mTextView.setText(ss);
    }

    @Override
    public int getItemCount() {
        return mlog.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    static class LogViewHolder extends RecyclerView.ViewHolder {
        final AppCompatTextView mTextView;

        LogViewHolder(AppCompatTextView view) {
            super(view);
            mTextView = view;
        }
    }
}

File: app/src/main/java/com/example/toucheventexplorer/logger/OnScreenLogAdapter.kt
package com.example.toucheventexplorer.logger

import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatTextView
import androidx.recyclerview.widget.RecyclerView
import javax.annotation.Nonnull

internal class OnScreenLogAdapter(@param:Nonnull private val mlog: List<EventLogEntry>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder?>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val line = AppCompatTextView(parent.context)
        line.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        return LogViewHolder(line)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as LogViewHolder
        viewHolder.mTextView.text = mlog[position].spannableString
    }

    override fun getItemCount() = mlog.size

    override fun getItemId(position: Int) = position.toLong()

    internal class LogViewHolder(val mTextView: AppCompatTextView) :
        RecyclerView.ViewHolder(mTextView)
}